lib/knapsack_pro/extensions/rspec_extension.rb

```diff
@@ -0,0 +1,164 @@
+        # Set with --error-exit-code CODE - Override the exit code used when there are errors loading or running specs outside of examples.
+        def knapsack__error_exit_code
+          _default = nil || configuration.error_exit_code
+        end
+
+        def knapsack__deprecated_run_all_when_everything_filtered_enabled?
+          raise "Cannot call the #{__method__} method because setup was not done." unless knapsack__setup_done?
+
+          !!(configuration.respond_to?(:run_all_when_everything_filtered) && configuration.run_all_when_everything_filtered)
+        end
+
+        def knapsack__seed
+          configuration.seed.to_s
+        end
+
+        def knapsack__seed_used?
+          configuration.seed_used?
+        end
+
+        # @param test_file_paths Array[String]
+        #   Example: ['a_spec.rb', 'b_spec.rb[1:1]']
+        def knapsack__load_spec_files_batch(test_file_paths)
+          world.reset
+
+          # Reset filters, but do not reset `configuration.static_config_filter_manager` to preserve the --tag option
+          filter_manager = RSpec::Core::FilterManager.new
+          options.configure_filter_manager(filter_manager)
+          configuration.filter_manager = filter_manager
+
+          configuration.knapsack__load_spec_files(test_file_paths)
+        end
+
+        # Based on:
+        # https://github.com/rspec/rspec-core/blob/f8c8880dabd8f0544a6f91d8d4c857c1bd8df903/lib/rspec/core/runner.rb#L113
+        #
+        # Option: `configuration.fail_if_no_examples`
+        #   Ignore the configuration.fail_if_no_examples option because it should be off in Queue Mode.
+        #   * Ignore the fail_if_no_examples option because in Queue Mode a late CI node can start after other CI nodes already executed tests. It is expected not to run examples in such a scenario.
+        #   * RSpec should not fail when examples are not executed for a batch of tests fetched from the Queue API. The batch could have tests that have no examples (for example, someone commented out the content of the spec file). We should fetch another batch of tests from the Queue API and keep running tests.
+        #
+        # @return [Fixnum] exit status code.
+        def knapsack__run_specs(queue_runner)
+          # Based on:
+          # https://github.com/rspec/rspec-core/blob/f8c8880dabd8f0544a6f91d8d4c857c1bd8df903/lib/rspec/core/world.rb#L53
+          ordering_strategy = configuration.ordering_registry.fetch(:global)
+          node_examples_passed = true
+
+          configuration.with_suite_hooks do
+            exit_status = configuration.reporter.report(_expected_example_count = 0) do |reporter|
+              queue_runner.with_batched_tests_from_queue do |test_file_paths|
+                knapsack__load_spec_files_batch(test_file_paths)
+
+                examples_passed = ordering_strategy.order(world.example_groups).map do |example_group|
+                  queue_runner.handle_signal!
```

Comment:
Yes. I added test to cover the summary of pending & failing tests is displayed as well.
https://github.com/KnapsackPro/knapsack_pro-ruby/pull/237/commits/488b096a860999b3072f30d259da82d21cc934e2

Reply:
I don't get it, is this story solved then? https://trello.com/c/wrIb78Yw/179-fix-ctrl-c-when-running-locally

Reply:
@shadre I think it's one of the many being solved 🙂

Reply:
In RSpec just one test example is executed after the INT signal is received and then process is shutdown:

```
...
  is expected to have db column named project_id
^C <--- here <ctrl-c> received
RSpec is shutting down and will print the summary report... Interrupt again to force quit.
  is expected to have db column named created_at

Finished in 0.6762 seconds (files took 2.73 seconds to load)
49 examples, 0 failures

Randomized with seed 58250
```

Currently, the knapsack_pro gem 6.0.3 on INT signal runs tests until the batch of tests fetched from the Queue API is executed. We made an improvement in this PR to terminate sooner after the top-level ExampleGroup is executed for a given test file (in practice after a single test file is executed unless it has a few `describe` blocks at top level of the file).


The only thing I was not able to solve in Queue Mode is to cancel after a single test example is executed as in RSpec. After we receive the INT signal the whole ExampleGroup (`describe` block) needs to finish running tests because we run it with `example_group.run(reporter)` in `lib/knapsack_pro/extensions/rspec_extension.rb` in `        def knapsack__run_specs(queue_runner)` method.

We would have to refactor the `example_group.run(reporter)` method to make termination after a single example instead of ExampleGroup. I did not expand the scope of this PR story so https://trello.com/c/wrIb78Yw/179-fix-ctrl-c-when-running-locally might still be relevant, but I'm not sure if it's needed.

Reply:
We can move that to done when this is merged 👍